### ISSUE TITLE
test_gzip_is_binary is failing

### ISSUE THREAD

**Issue by grawlinson** [Wilfred/difftastic#478]
Hey! Just updating this for Arch Linux (`0.42.0` to `0.43.0`), and one test in particular is failing:

```sh
failures:

---- files::tests::test_gzip_is_binary stdout ----
thread 'files::tests::test_gzip_is_binary' panicked at 'assertion failed: `(left == right)`
  left: `Text("謟\u{8}뀺掑\u{300}躋˥䐀棒Ͱ\0")`,
 right: `Binary`', src/files.rs:299:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    files::tests::test_gzip_is_binary

test result: FAILED. 117 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s
```

I think the offender is related to PR #473.

**Comment by grawlinson** [comment]
Here's a full backtrace, if that helps.

<details><summary>full backtrace</summary>

```sh
test files::tests::test_gzip_is_binary ... FAILED

failures:

---- files::tests::test_gzip_is_binary stdout ----
thread 'files::tests::test_gzip_is_binary' panicked at 'assertion failed: `(left == right)`
  left: `Text("謟\u{8}뀺掑\u{300}躋˥䐀棒Ͱ\0")`,
 right: `Binary`', src/files.rs:299:9
stack backtrace:
   0:     0x55656304087a - std::backtrace_rs::backtrace::libunwind::trace::h71fbe62a666bb430
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x55656304087a - std::backtrace_rs::backtrace::trace_unsynchronized::h225783a6d786a86c
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x55656304087a - std::sys_common::backtrace::_print_fmt::hbc5c5ab7371a7243
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x55656304087a - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hc51bcaab7a8c817d
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x55656306448e - core::fmt::write::h1afbf121bfdf9387
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/core/src/fmt/mod.rs:1208:17
   5:     0x55656303d565 - std::io::Write::write_fmt::h18c9a8b99e7c546c
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/io/mod.rs:1682:15
   6:     0x556563040645 - std::sys_common::backtrace::_print::h16fce32ac6eeb2d5
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/sys_common/backtrace.rs:47:5
   7:     0x556563040645 - std::sys_common::backtrace::print::h0701d6b944b71665
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/sys_common/backtrace.rs:34:9
   8:     0x5565630422bf - std::panicking::default_hook::{{closure}}::hcc16958146635e82
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/panicking.rs:267:22
   9:     0x556563041f7f - std::panicking::default_hook::he61064dc7d426de3
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/panicking.rs:283:9
  10:     0x556562d595c7 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hfd0faf7e1a55b1d2
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/alloc/src/boxed.rs:2032:9
  11:     0x556562d595c7 - test::test_main::{{closure}}::hdadf1fe9b5fac34f
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/test/src/lib.rs:135:21
  12:     0x5565630429ed - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hc7b520dbafff85be
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/alloc/src/boxed.rs:2032:9
  13:     0x5565630429ed - std::panicking::rust_panic_with_hook::hd139f9626cd45b6e
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/panicking.rs:692:13
  14:     0x556563042769 - std::panicking::begin_panic_handler::{{closure}}::h100239264035d01d
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/panicking.rs:579:13
  15:     0x556563040d2c - std::sys_common::backtrace::__rust_end_short_backtrace::h00ce37c67437ff87
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/sys_common/backtrace.rs:137:18
  16:     0x556563042472 - rust_begin_unwind
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/panicking.rs:575:5
  17:     0x5565627463b3 - core::panicking::panic_fmt::h478fac9a730af6b7
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/core/src/panicking.rs:64:14
  18:     0x5565627466cf - core::panicking::assert_failed_inner::h711985871cdd2d28
  19:     0x5565627ae56a - core::panicking::assert_failed::h5a173bcb22c2943d
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/core/src/panicking.rs:199:5
  20:     0x556562815dbd - difft::files::tests::test_gzip_is_binary::h69498132e7f149d0
                               at /build/difftastic/src/difftastic/src/files.rs:299:9
  21:     0x55656286d5b7 - difft::files::tests::test_gzip_is_binary::{{closure}}::ha762f53329baec60
                               at /build/difftastic/src/difftastic/src/files.rs:292:30
  22:     0x55656281a895 - core::ops::function::FnOnce::call_once::h87e2bc0495cbe65f
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/core/src/ops/function.rs:507:5
  23:     0x556562d5e9ff - core::ops::function::FnOnce::call_once::h2cf0ec6dd87f8b7c
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/core/src/ops/function.rs:507:5
  24:     0x556562d5e9ff - test::__rust_begin_short_backtrace::h6bd239829b349205
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/test/src/lib.rs:657:18
  25:     0x556562d301ec - test::run_test::{{closure}}::ha56f0fbcdb0f996b
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/test/src/lib.rs:648:30
  26:     0x556562d301ec - core::ops::function::FnOnce::call_once{{vtable.shim}}::h8c26456a1d7a0321
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/core/src/ops/function.rs:507:5
  27:     0x556562d5d9ea - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::ha827cdd90f9dbb41
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/alloc/src/boxed.rs:2000:9
  28:     0x556562d5d9ea - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h10faf1853f836a83
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/core/src/panic/unwind_safe.rs:271:9
  29:     0x556562d5d9ea - std::panicking::try::do_call::h00a91caa306c6634
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/panicking.rs:483:40
  30:     0x556562d5d9ea - std::panicking::try::h2a755cf6671db02d
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/panicking.rs:447:19
  31:     0x556562d5d9ea - std::panic::catch_unwind::h08ad77293f1f8a5e
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/panic.rs:137:14
  32:     0x556562d5d9ea - test::run_test_in_process::h1aa106702f6865cb
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/test/src/lib.rs:680:27
  33:     0x556562d5d9ea - test::run_test::run_test_inner::{{closure}}::hdfd143275a5cc26c
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/test/src/lib.rs:574:39
  34:     0x556562d2a941 - test::run_test::run_test_inner::{{closure}}::h17c7a672cc5983a0
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/test/src/lib.rs:601:37
  35:     0x556562d2a941 - std::sys_common::backtrace::__rust_begin_short_backtrace::h119312fee89fe50e
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/sys_common/backtrace.rs:121:18
  36:     0x556562d3027b - std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}::h4493630c5eb42471
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/thread/mod.rs:550:17
  37:     0x556562d3027b - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h109e0534f8cd4b6c
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/core/src/panic/unwind_safe.rs:271:9
  38:     0x556562d3027b - std::panicking::try::do_call::h154b8e0bdc7cac70
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/panicking.rs:483:40
  39:     0x556562d3027b - std::panicking::try::h8bbe468791d19f37
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/panicking.rs:447:19
  40:     0x556562d3027b - std::panic::catch_unwind::hf4cbcc8b525bca3a
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/panic.rs:137:14
  41:     0x556562d3027b - std::thread::Builder::spawn_unchecked_::{{closure}}::hb310eceab28460af
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/thread/mod.rs:549:30
  42:     0x556562d3027b - core::ops::function::FnOnce::call_once{{vtable.shim}}::hb1378dd3f8f8a8d1
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/core/src/ops/function.rs:507:5
  43:     0x556563046f63 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h5d16a47467e7e41d
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/alloc/src/boxed.rs:2000:9
  44:     0x556563046f63 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h962f776b429f07a4
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/alloc/src/boxed.rs:2000:9
  45:     0x556563046f63 - std::sys::unix::thread::Thread::new::thread_start::h24c161f0c800f863
                               at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/sys/unix/thread.rs:108:17
  46:     0x7fa1601b78fd - <unknown>
  47:     0x7fa160239d20 - <unknown>
  48:                0x0 - <unknown>


failures:
    files::tests::test_gzip_is_binary

test result: FAILED. 117 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.03s

error: test failed, to rerun pass `--bin difft`
```
</details>

**Comment by Wilfred** [comment]
Argh, looks like the test fails when there isn't a MIME type database in the current environment.

**Comment by icp1994** [comment]
same happened while updating in void linux: https://github.com/void-linux/void-packages/pull/42081

**Comment by grawlinson** [comment]
Do you mean a package like [this](https://archlinux.org/packages/extra/x86_64/shared-mime-info/)? I’m not at the computer at the moment, but I’ll rerun the tests with this installed and see if the test passes.

**Comment by Wilfred** [comment]
Yep, I'd consider the mime database to be an optional dependency, but strongly recommended. From a packaging perspective it might be better to consider it required: you get better file detection (e.g. gzip) if you have it installed.

There's a note in the manual here about the database: https://difftastic.wilfred.me.uk/installation.html#optional-install-mime-database

I agree the test should just be skipped by default, so 34f21c6d9f256f16cc5f77294bcca9e61cee3fd6 does that :)

**Comment by Wilfred** [comment]
I've released 0.43.1, so this should now be fixed :) https://github.com/Wilfred/difftastic/releases/tag/0.43.1

**Comment by grawlinson** [comment]
Adding `shared-mime-info` as a test dependency results in the test passing, so I've just added it as an explicit run-time dependency. Additionally, all ignored tests are run in our testing environment.

BTW `--ignored` seems to *only* run ignored tests, according to the [CI](https://github.com/Wilfred/difftastic/actions/runs/4093652298/jobs/7059244036#step:4:176). Might have to run tests twice like so:

```sh
# run all tests (excluding ignored)
cargo test
# run ignored tests
cargo test -- --ignored
```
